Add packaging support for Docker images

bahnwaerter opened the pull request:
The Docker packaging support allows to build Docker images for all Ultimate tools. The images are based on Alpine Linux to ensure a minimal footprint. The build is based on a multi-stage process where temporary build artifacts are discarded and not included in the final Docker images. Each Ultimate tool now has its own dedicated Docker image. Additionally, the WebBackend can be automatically configured through docker-compose environment files to simplify the configuration and deployment of the Web service.

bahnwaerter committed:
Build and package Ultimate for Docker

bahnwaerter committed:
Fix newline escaping issues in Dockerfile

bahnwaerter committed:
Fix config issues regarding dockerize and worng paths

bahnwaerter committed:
Remove broken dockerize waiting until service is ready

bahnwaerter committed:
Optimize multi-stage Docker build for all Ultimate applications

bahnwaerter committed:
Update Debian in Docker images to Debian Bookworm

bahnwaerter committed:
Execute Ultimate as non-root user and migrate Docker packaging to Alpine Linux

bahnwaerter commented:
Currently, the Ultimate tools are built with individual build calls, and the appropriate configurations (toolchain and settings files) are still missing in the respective Docker images.

It would be advantageous if the complete build of the Docker images would be based on the `makeFresh.sh` script. However, this also includes the `makeZip.sh` functionality, which is not necessarily required here. Should we split the functionality of the two scripts with regard to the Docker packaging?

A new separation of functionality could look like this:
- `makeBuild.sh`: Responsible for the Maven build and to collect the correct configurations for each tool
- `makeZip.sh`: Compresses the corresponding built artifacts
- `makeFresh.sh`: Calls `makeBuild.sh` and then `makeZip.sh`

The separation introduces the `makeBuild.sh` script that could be used to build the Docker images.

bahnwaerter committed:
Add Docker image for Ultimate's web frontend based on Jekyll

danieldietsch commented:
That's a very good idea. I also like your Dockerfile, in particular the use of dockerize!

bahnwaerter commented on packaging/docker/Dockerfile:
```diff
@@ -44,6 +46,40 @@ RUN mkdir -p "${DIR_BUILD}" && \
+RUN echo -e "gem 'csv'\n" \
+            "gem 'base64'\n" \
+            "gem 'bigdecimal'\n" \
+            "gem 'logger'" >> Gemfile
```
As soon as the reported issue #740 is fixed, this patch can be removed.

bahnwaerter replied:
The issue reported in #740 is fixed and the patch can be removed!

bahnwaerter commented on packaging/docker/docker-compose.yml:
```diff
@@ -0,0 +1,25 @@
+    image: "ultimate-webbackend"
+    env_file: "ultimate-webbackend.env"
+    ports:
+      - "8080:8080"
```
The port of the backend should not be hard-coded here.
Instead, the value from the Docker Compose environment file should be used.

bahnwaerter commented on packaging/docker/docker-compose.yml:
```diff
@@ -0,0 +1,25 @@
+    image: "ultimate-webfrontend"
+    env_file: "ultimate-webfrontend.env"
+    ports:
+      - "80:80"
```
The port of the frontend should not be hard-coded here.
Instead, the value from the Docker Compose environment file should be used.

bahnwaerter commented on packaging/docker/Dockerfile:
```diff
@@ -0,0 +1,388 @@
+            "gem 'logger'" >> Gemfile
+
+ENV PATH "${DIR_OUTPUT}/CLI-E4/linux/gtk/x86_64:${PATH}"
+RUN python3 scripts/build.py
```
The build cannot be configured by passing appropriate values, such as setting an arbitrary `baseurl` for Jekyll.
The used Python scripts must be adapted accordingly to make the Jekyll build configurable.

maul-esel replied:
There is a `--production` parameter that, among other things, sets the `baseurl`. But if you need more flexibility, then the script can be made more flexible, yes.

bahnwaerter replied:
I've already discovered the production parameter. This paraemter sets the `baseurl` to the fixed value `/`. Therefore, we should expand the script slightly to allow any `baseurl`  for the Docker image build.

bahnwaerter commented on packaging/docker/Dockerfile:
```diff
@@ -0,0 +1,388 @@
+RUN ln -s "${USER_HOME}/UltimateDebug" "/usr/bin/UltimateDebug"
+
+# run the Ultimate Debug UI product as non-root user
+#USER "${USER_NAME}"
```
Try running the debug UI in a foreign X session as a regular user within the container instead of using the root user.

bahnwaerter committed:
Remove patch for Jekyll in Docker Web frontend image build

The patch for the Jekyll installation is no longer necessary since the
issue was fixed by updating Jekyll in commit c0f9255.

bahnwaerter committed:
Run the Ultimate Debug UI in a Docker container as non-root user

bahnwaerter committed:
Make Jekyll baseurl option configurable for a Docker image build

bahnwaerter committed:
Make Docker Compose setup of the Ultimate Web service configurable

bahnwaerter committed:
Separate build of Ultimate products from creating ZIP archives

This change introduces a new makeBuild.sh script that builds all
Ultimate products and packages the products with the required
configurations (toolchain and settings files), additional programs
and files like solvers or licenses. This script is then used in the
existing makeFresh.sh script for the common build steps to avoid
duplicate script code. The functionality of the makeFresh.sh script
remains unchanged.

bahnwaerter committed:
Build and package Ultimate Eliminator as product

bahnwaerter committed:
Build and package Ultimate CLI and Debug-UI

bahnwaerter committed:
Use build script to create Docker images with product-specific config

bahnwaerter committed:
Indicate archive step after build with separate heading

bahnwaerter committed:
Add parameter to build script to restrict platforms

bahnwaerter committed:
Add notes and restrictions to Docker packaging README

bahnwaerter committed:
Fix Docker run command for Ultimate images

bahnwaerter committed:
Replace Perl Regex with Extended Regex to support non-GNU grep

bahnwaerter committed:
Replace GNU readlink in build scripts

bahnwaerter committed:
Fix quoting in build scripts

bahnwaerter committed:
Fix invalid path to Ultimate product in Dockerfile

bahnwaerter committed:
Enable colorized promt in Docker container

bahnwaerter committed:
Make Ultimate executable, scripts, and solvers in Docker image available

bahnwaerter committed:
Adjust Ultimate configuration path variable in Docker container

bahnwaerter committed:
Make maximum heap and stack size configurable for Ultimate products

bahnwaerter committed:
Fix invalid Docker compose build argument for the Webbackend image

bahnwaerter committed:
Make maximum heap and stack size configurable in Docker compose setup

bahnwaerter committed:
Add GCC as syntax checker to Docker image of WebBackend

bahnwaerter committed:
Fix invalid Docker defaults for the Ultimate Webservices

danieldietsch commented on packaging/docker/Dockerfile:
```diff
@@ -0,0 +1,516 @@
+
+# build all Ultimate products
+WORKDIR "${DIR_ULTIMATE}"
+RUN cd "${DIR_OUTPUT_BUILD}" && bash makeBuild.sh -m "${ULTIMATE_MEMORY_HEAP_MAX}" -s "${ULTIMATE_MEMORY_STACK_MAX}" -p linux
```
Would it make sense to also support the other platforms? In particular Mac might be interesting to some, but I guess we still don't have anybody who could test that. 
Instead, can you add a line in the README that says why it is limited to Linux?

bahnwaerter replied:
> Would it make sense to also support the other platforms?
> Instead, can you add a line in the README that says why it is limited to Linux?

Sure, that would be a great idea. We could especially add packaging support for the already built MacOS version.
Note that the restriction in the `makeBuild.sh` script is only used for building the Docker images and only affects the final packaging, not the Maven build. I've already added information to the `README.md` that we currently only support Linux containers for Docker. These can also be run on Windows using Docker Desktop with the WSL2 backend. Building native Windows containers for Docker is pretty annoying, but not impossible. Furthermore, I doubt users will want to deal with the licensing issue of Windows containers.

danieldietsch commented on packaging/docker/Dockerfile:
```diff
@@ -0,0 +1,516 @@
+                       "ruby-dev"
+
+# install 'dockerize' for automatic configuration file creation
+ENV DOCKERIZE_VERSION="v0.9.5"
```
I would place this ENV together with the others at the beginning

danieldietsch commented on packaging/docker/Dockerfile:
```diff
@@ -0,0 +1,516 @@
+ARG DIR_TOOLS
+ARG USER_NAME
+ARG USER_GROUP
+ARG USER_HOME
```
Are all of these args necessary for the packaging stages? I dont quite remember if you need to redeclare them for the dependent images...

bahnwaerter replied:
Yes, the re-declaration of shared build `ARG`s are necessary in a multi-stage Docker file since they only last for the build phase of a single Docker image.

danieldietsch commented on packaging/docker/Dockerfile:
```diff
@@ -0,0 +1,516 @@
+# default entry point to serve the Ultimate web frontend
+ENTRYPOINT dockerize -template "config.js.tpl:js/webinterface/config.js" \
+                     bundle exec jekyll serve --host "${ULTIMATE_FRONTEND_HOST}" \
+                                              --port "${ULTIMATE_FRONTEND_PORT}"
```
Hm, why not build the frontend with jekyll and serve it with a "real" webserver? I was at least under the impression that ``jekyll serve`` is only used for debugging/designing a website, not really for production use.

bahnwaerter replied:
I created the Docker image of the Web frontend for development purposes. Here, `jekyll serve` offers an automatic rebuild and reload mechanism, which makes it super easy to debug and extend the frontend.
This is obviously not suitable for a productive Docker setup. In this case, you would probably be better off using a decent Web server, like your proposed nginx, and a powerful proxy server, like Traefik.

We could create a second Docker image of the Web frontend for production purposes. What do you think?

danieldietsch replied:
Ah, that makes sense. Perhaps a comment in the README would be nice so that we remember why it is like it is :)

And of course,  a second variant with nginx would be appreciated, then I would switch our current setup over to that.

danieldietsch commented on packaging/docker/Dockerfile:
```diff
@@ -0,0 +1,516 @@
+# Serving the frontend from the backend is not supported by this Docker image
+ENV ULTIMATE_FRONTEND_SERVE="False"
+ENV ULTIMATE_FRONTEND_PATH="${USER_HOME}"
+ENV ULTIMATE_FRONTEND_ROUTE="/website"
```
Again, I would move all ENVs to the front if they are meant to be editable by someone.

bahnwaerter replied:
These settings are actually hidden because they are not meant to be editable, but rather to deactivate the Web frontend serving of the Jetty web service. Unfortunately, these variables cannot be static Docker build `ARG`s, as otherwise `dockerize` can no longer use them.

danieldietsch commented on packaging/docker/README.md:
```diff
@@ -0,0 +1,79 @@
+> [!NOTE]
+> Building the Ultimate product images is currently limited to the Docker target platform `linux/amd64` (Linux containers for the 64-bit x86 architecture).
+> However, these images can still be used on a Windows system with [Docker Desktop](https://docs.docker.com/desktop/setup/install/windows-install/) configured with the WSL2 or Hyper-V backend in order to create and run Linux containers on a Windows system.
+
```
Thats the place for the reason :)

bahnwaerter replied:
Is the explanation sufficient for you?

danieldietsch replied:
No :) -- that does not contain the reason you gave in the other discussion, and it would be nice to have it here.

danieldietsch commented on packaging/docker/README.md:
```diff
@@ -0,0 +1,79 @@
+Calling the Ultimate `PRODUCT` within the container then follows as usual, where a `TOOLCHAIN`, `SETTINGS`, and `PROGRAM` file should be specified for a verification run.
+
+> [!NOTE]
+> The pre-configured products are already provided with the appropriate configuration (toolchain and setting files).
```
Confusingly worded: does this mean I dont have to specify them in the run command above?

bahnwaerter replied:
You are right. This is very badly written. I will rephrase the note.

> [...] does this mean I dont have to specify them in the run command above?

No, that is not what the note means :rofl: What is meant here is that the images for the containers already ship the product-specific configuration files needed to call the Ultimate `PRODUCT`.

danieldietsch commented on packaging/docker/docker-compose.yml:
```diff
@@ -0,0 +1,49 @@
+version: "2"
```
Recent docker-compose files do not have the version anymore

danieldietsch commented on packaging/docker/welcome.sh:
```diff
@@ -0,0 +1,34 @@
+echo "▗▖ ▗▖▗▖ ▗▄▄▄▖▗▄▄▄▖▗▖  ▗▖ ▗▄▖▗▄▄▄▖▗▄▄▄▖"
+echo "▐▌ ▐▌▐▌   █    █  ▐▛▚▞▜▌▐▌ ▐▌ █  ▐▌   "
+echo "▐▌ ▐▌▐▌   █    █  ▐▌  ▐▌▐▛▀▜▌ █  ▐▛▀▀▘"
+echo "▝▚▄▞▘▐▙▄▄▖█  ▗▄█▄▖▐▌  ▐▌▐▌ ▐▌ █  ▐▙▄▄▖"
```
🤣

danieldietsch commented on packaging/docker/welcome.sh:
```diff
@@ -0,0 +1,34 @@
+        if command -v "${cmd}" >/dev/null 2>&1; then
+            "${cmd}" --version
+            return
+        fi
```
else?

bahnwaerter replied:
The `else` branch was intentionally omitted because the script is very general and intended to print version information for each Ultimate product. Therefore, the script checks all known launcher names and prints the version information for the first one that exists in the Docker container.

danieldietsch replied:
Ah, that makes sense -- I was afraid that this leads to silent failures when no command matches

danieldietsch commented on releaseScripts/default/adds/Ultimate.ini:
Is this not needed anymore?

bahnwaerter replied:
Yes, this configuration file is no longer needed. It is automatically generated from the arguments specified in the `*.product` files. At the end, the package script adjusts the memory sizes for the Java Runtime as needed.

bahnwaerter replied:
One advantage of this approach is that the launcher settings are always adjusted automatically and do not have to be changed after framework migrations.

danieldietsch commented on releaseScripts/default/makeBuild.sh:
```diff
@@ -0,0 +1,193 @@
+
+_print_help() {
+  printf 'Usage: %s [-m <size>] [-s <size>] [-p all|linux|win32] [-h]\n' "${0}"
+  print_newline
```
feels overengineered ;) -- why not `echo`?

bahnwaerter replied:
That's probably in the nature of a C programmer :wink:

danieldietsch replied:
:+1:

danieldietsch commented on releaseScripts/default/makeZip.sh:
I think this could now become a function in makeFresh -- which could also be somewhat simpler then. In particular, now that packaging is done before and this only names the zip and actually creates it.

bahnwaerter replied:
Good idea! The script has become very small now. It's really not worth a script anymore.

danieldietsch commented on trunk/source/BA_SiteRepository/pom.xml:
```diff
@@ -54,7 +54,7 @@
 								</product>
 								<product>
-									<id>UltimateEliminator</id>
+									<id>Eliminator</id>
```
Why remove the Ultimate prefix here?

bahnwaerter replied:
This will finally give us a consistent naming scheme. The actual product name simplifies the folder naming for the packaging process and keeps the packaging logic generic. Until now, this product name has always been out of place. Not even the Delta Debugger beat that :wink:

danieldietsch replied:
Fair

danieldietsch commented on trunk/source/CoreRCP/plugin.xml:
```diff
@@ -37,8 +37,8 @@
    
-   <extension point="org.eclipse.core.runtime.products" id="UltimateEliminator">
-      <product application="de.uni_freiburg.informatik.ultimate.core.CoreRCP" name="UltimateEliminator">
+   <extension point="org.eclipse.core.runtime.products" id="Eliminator">
```
see above

danieldietsch commented on trunk/source/WebsiteStatic/scripts/build.py:
```diff
@@ -28,16 +28,24 @@ def parse_args() -> argparse.Namespace:
             "--skip-settings", action="store_true", help="skip rebuilding settings (useful during debugging)"
         )
+        parser.add_argument(
+            "--baseurl", action="store", type=str, help="set base URL for site (necessary when hosting the site in a sub-directory)"
```
default is `website` right?

bahnwaerter replied:
Yes. This value is everywhere (Docker file, Docker Compose) the default value.

danieldietsch reviewed (commented):
Very cool changes! 

I added some minor comments, and have two things:

#### Use "real" webserver
I would like to avoid using the Jekyll webserver for the frontend. Could you just use nginx as base?

Our current frontend at https://ultimate-pa.org does the following.
```
services:
  frontend:
    image: nginx:latest
    volumes:
      - /mnt/shared/ultimate-pa/config/nginx/conf.d:/etc/nginx/conf.d
      - /mnt/shared/ultimate-pa/config/nginx/nginx.conf:/etc/nginx/nginx.conf
      - /mnt/shared/ultimate-pa/frontend:/www/frontend
      - /mnt/shared/ultimate-pa/config/frontend/config.js:/www/frontend/config/config.j
```
where ``/mnt/shared/ultimate-pa/frontend`` is the static website, ``config.js`` the config you would expect, ``/mnt/shared/ultimate-pa/config/nginx/nginx.conf`` is default, and `/mnt/shared/ultimate-pa/config/nginx/conf.d/default.conf` is default except specifying the frontend folder:
```
server {
    listen       80;
    listen  [::]:80;
    server_name  localhost;

    location / {
        root   /www/frontend;
        index  index.html index.htm;
    }

    error_page   500 502 503 504  /50x.html;
    location = /50x.html {
        root   /usr/share/nginx/html;
    }

    error_page  404 /404.html;
    location /404.html {
        root /www/frontend;
        internal;
    }
}
```

#### Check windows build
I tried running ``./makeFresh.sh`` as usual, and got at the end:
```
Archive Ultimate Debug UI [linux]
━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
Not enough arguments supplied -- use arguments in the following order
1. the toolname
2. the launcher name
3. the maximum heap memory size
4. the maximum stack memory size
5. 'linux' or 'win32' for the target platform
```
But everything else worked as expected

bahnwaerter committed:
Fix issue when archiving the Ultimate Debug UI product

bahnwaerter committed:
Make maximum heap and stack size configurable in Ultimate.py

bahnwaerter committed:
Fix invalid build argument in Docker compose setup

bahnwaerter committed:
Fix incorrect handling of memory settings for Ultimate products

bahnwaerter committed:
Make maximum heap and stack size configurable in run_complete_analysis.py

bahnwaerter committed:
Change the default memory sizes of the Java Runtime for Ultimate products

The standard memory sizes of the Java Runtime for Ultimate products are
uniformly set to the following values:

    - Initial heap  memory size: 2M
    - Maximum heap  memory size: 4G
    - Maximum stack memory size: 1M

If other memory sizes are required, these can be configured for each
build of Ultimate products. The Ultimate products for the SV-COMP should
be built with the following settings:

    ./makeBuild.sh -i 2M -m 15G -s 1M

On the other hand, the ReqCheck product for analyzing large requirement
sets should be built with the following settings:

    ./makeBuild.sh -i 2M -m 100G -s 4M

schuessf commented on releaseScripts/default/semver2.sh:
```diff
@@ -0,0 +1,354 @@
+  printf %s "$next"
+}
+
+semver_compare() {
```
I just tried running `makeFresh.sh`, but it just showed the build splashscreen and exited silently.
I checked and it turned out that `javac --version` produced `javac 24`, which should be >= 21.
After I used `update-alternatives` to change javac to 21, the build worked as expected.
Therefore, I think there is some issue here.

bahnwaerter replied:
You're right. The version comparison should support that.
